components/telemetry-operator/internal/sync/sync.go

```diff
@@ -257,56 +258,100 @@ func (s *LogPipelineSyncer) syncFilesConfigMap(ctx context.Context, logPipeline
 	return changed, nil
 }
 
-// Copy referenced secrets to global Fluent Bit environment secret.
-func (s *LogPipelineSyncer) syncSecretRefs(ctx context.Context, logPipeline *telemetryv1alpha1.LogPipeline) (bool, error) {
+func isSecretRef(fromType telemetryv1alpha1.ValueFromType) bool {
+	return fromType.SecretKey.Name != "" && fromType.SecretKey.Key != ""
+}
+
+func (s *LogPipelineSyncer) fetchSecret(ctx context.Context, fromType telemetryv1alpha1.ValueFromType) (*corev1.Secret, error) {
 	log := logf.FromContext(ctx)
-	secret, err := s.getOrCreateSecret(ctx, s.DaemonSetConfig.FluentBitEnvSecret)
+
+	secretKey := fromType.SecretKey
+	var referencedSecret corev1.Secret
+	if err := s.Get(ctx, types.NamespacedName{Name: secretKey.Name, Namespace: secretKey.Namespace}, &referencedSecret); err != nil {
+		log.Error(err, "Failed reading secret %s from namespace %s", secretKey.Name, secretKey.Namespace)
+		return nil, err
+	}
+	return &referencedSecret, nil
+
+}
+
+// syncVariables copies referenced secrets to global Fluent Bit environment secret.
+func (s *LogPipelineSyncer) syncVariables(ctx context.Context, logPipeline *telemetryv1alpha1.LogPipeline) (bool, error) {
+	log := logf.FromContext(ctx)
+	oldSecret, err := s.getOrCreateSecret(ctx, s.DaemonSetConfig.FluentBitEnvSecret)
 	if err != nil {
 		return false, err
 	}
 
-	changed := false
-	for _, secretRef := range logPipeline.Spec.SecretRefs {
-		var referencedSecret corev1.Secret
-		if err = s.Get(ctx, types.NamespacedName{Name: secretRef.Name, Namespace: secretRef.Namespace}, &referencedSecret); err != nil {
-			log.Error(err, "Failed reading secret %s from namespace %s", secretRef.Name, secretRef.Namespace)
-			continue
-		}
-		for k, v := range referencedSecret.Data {
-			if logPipeline.DeletionTimestamp != nil {
-				if _, hasKey := secret.Data[k]; hasKey {
-					delete(secret.Data, k)
-					controllerutil.RemoveFinalizer(logPipeline, secretRefsFinalizer)
-					changed = true
+	newSecret := oldSecret
+	newSecret.Data = make(map[string][]byte)
+
+	var logPipelines telemetryv1alpha1.LogPipelineList
+	err = s.List(ctx, &logPipelines)
+	if err != nil {
+		return false, err
+	}
+
+	if logPipeline.DeletionTimestamp != nil {
+		controllerutil.RemoveFinalizer(logPipeline, secretRefsFinalizer)
+	}
+
+	for _, l := range logPipelines.Items {
+		for _, varRef := range l.Spec.Variables {
```

Comment:
We should skip LogPipelines that have a DeletionTimestamp. Would probably not make a big difference since they will finally disappear and not be added again in the next reconciliation round, but skipping them would immediately result in the desired state.

Reply:
I fetch the current list of logpipelines and here we would not have the deleting logpipeline. So thats why have moved the logic related to deleting log pipeline outside.